File: 디자인패턴/Python 디자인 패턴/Mediator/SmartHome.py
from Mediator import*

from Door import*
from Window import*
from CoolAircon import*
from HeatBoiler import*

class SmartHome (Mediator):
    def __init__(this) -> None:
        this.PartiMedi = {
            "door":Door(this),
            "window":Window(this),
            "aircon":CoolAircon(this),
            "boiler":HeatBoiler(this)
        }
    
    def participantChange(this, participant:Participant)->None:
        if((participant==this.PartiMedi["door"] and
            this.PartiMedi["door"].isDoor())
            or (participant==this.PartiMedi["window"] and
            this.PartiMedi["window"].isWindow())):
            this.PartiMedi["aircon"].Off()
            this.PartiMedi["boiler"].Off()
        elif(participant==this.PartiMedi["aircon"] and
            this.PartiMedi["aircon"].isAircon()):
            this.PartiMedi["door"].Close()
            this.PartiMedi["window"].Close()
            this.PartiMedi["boiler"].Off()
        elif(participant==this.PartiMedi["boiler"] and
            this.PartiMedi["boiler"].isBoiler()):
            this.PartiMedi["door"].Close()
            this.PartiMedi["window"].Close()
            this.PartiMedi["aircon"].Off()
    
    def report(this):
        print("[집안 상태]")
        print(this.PartiMedi["door"].strPrint())
        print(this.PartiMedi["window"].strPrint())
        print(this.PartiMedi["aircon"].strPrint())
        print(this.PartiMedi["boiler"].strPrint())
        print()
    pass

File: 디자인패턴/java 디자인 패턴/Mediator/SmartHome.java
public class SmartHome implements Mediator {
    public Door door = new Door(this);
    public Window window = new Window(this);
    public CoolAircon aircond = new CoolAircon(this);
    public HeatBoiler boiler = new HeatBoiler(this);


    @Override
    public void participantChange(Participant participant) {
        if((participant == door && door.isDoor()) || 
                (participant == window && window.isWindow())) {
            aircond.Off();
            boiler.Off();
        }
        else if(participant == aircond && aircond.isRunning()) {
            boiler.Off();
            window.Close();
            door.Close();
        }
        else if(participant == boiler && boiler.isRunning()) {
            aircond.Off();
            window.Close();
            door.Close();
        }
    }

    public void report() {
        System.out.println("[집안 상태]");
        System.out.println(door);
        System.out.println(window);
        System.out.println(aircond);
        System.out.println(boiler);
        System.out.println();
    }
}